ogr/ogrsf_frmts/flatgeobuf/ogr_flatgeobuf.h

```diff
@@ -97,7 +97,7 @@ class OGRFlatGeobufLayer final : public OGRLayer, public OGRFlatGeobufBaseLayerI
 
         // creation
         bool m_create = false;
-        std::vector<std::shared_ptr<FlatGeobuf::Item>> m_featureItems; // feature item description used to create spatial index
+        std::vector<FeatureItem> m_featureItems; // feature item description used to create spatial index
```

Comment:
How is this vector used?  Are these side effects intented?
- Adjacent items (wrt index in vector) will now be adjacent in memory. This can significantly improve certain access patterns. But
- Each slot in the vector now occupies the size of `FeatureItem` instead of the size of `std::shared_ptr`.
- Reallocations while appending to the vector will now copy `FeatureItem`s instead of smart pointers.

Reply:
Your points are valid. "Ideally" we'd probably want a some_container<std::vector<FeatureItem>> where the inner std::vector<> is limited to batches of let's say 1 million features to avoid too big memory reallocations, but that would be an implementation complication I'm not motivated enough to deal with.
In practice my change seems to work reasonably well, reducing memory usage and execution time, for the use case I tested (3.3 million features). It is hard to predict  which of the current/old strategy or the one of this PR would be better when dealing with much more features (100's of millions or more)

Before
---------

```
$ /usr/bin/time --verbose ogr2ogr out.fgb nz-building-outlines.fgb
	Command being timed: "ogr2ogr out.fgb nz-building-outlines.fgb"
	User time (seconds): 23.54
	System time (seconds): 3.25
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:26.90
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 489668
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 11
	Minor (reclaiming a frame) page faults: 158929
	Voluntary context switches: 215
	Involuntary context switches: 679
	Swaps: 0
	File system inputs: 1456
	File system outputs: 6857352
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Extract of ``cat /proc/$(pidof ogr2ogr)/status`` just before process terminates:
```
VmPeak:	  607968 kB
VmSize:	  580204 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	  488668 kB
VmRSS:	  459888 kB
RssAnon:	  428536 kB
RssFile:	   31352 kB
RssShmem:	       0 kB
VmData:	  441776 kB
VmStk:	     136 kB
VmExe:	      12 kB
VmLib:	   65528 kB
VmPTE:	    1080 kB
VmSwap:	       0 kB
```

Output of massif-visualizer on the file generated by ``valgrind --tool=massif ogr2ogr out.fgb nz-building-outlines.fgb``:
![image](https://user-images.githubusercontent.com/1192433/204091904-abd386dd-c341-4fd4-8dce-eb02ed0cb95a.png)

After
------

```
$ /usr/bin/time --verbose ogr2ogr out.fgb nz-building-outlines.fgb
	Command being timed: "ogr2ogr out.fgb nz-building-outlines.fgb"
	User time (seconds): 19.09
	System time (seconds): 3.01
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:22.12
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 359356
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 167308
	Voluntary context switches: 185
	Involuntary context switches: 112
	Swaps: 0
	File system inputs: 0
	File system outputs: 6857352
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

Extract of ``cat /proc/$(pidof ogr2ogr)/status`` just before process terminates:
```
VmPeak:	  511888 kB
VmSize:	  483856 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	  359164 kB
VmRSS:	  330168 kB
RssAnon:	  298068 kB
RssFile:	   32100 kB
RssShmem:	       0 kB
VmData:	  345440 kB
VmStk:	     136 kB
VmExe:	      12 kB
VmLib:	   65516 kB
VmPTE:	     828 kB
VmSwap:	       0 kB
```


Output of massif-visualizer on the file generated by ``valgrind --tool=massif ogr2ogr out.fgb nz-building-outlines.fgb``:
![image](https://user-images.githubusercontent.com/1192433/204091941-1ce479f1-2e5c-4569-ba87-b1dbb2a025c3.png)

I'm not clear why there are phases where memory usage decrease. Perhaps during the std::vector<> reallocation where you allocate 2 * old_size before releasing old_size. But one should also see that in the Before, perhaps to a lesser extent

Reply:
Measuring is great and beats all theory. : +1:

Just for completeness,  `std::unique_ptr` would give the smart pointer behaviour without  the specific overheads of `shared_ptr`. And `make_shared` would improve allocations of `std::shared.`

> "Ideally" we'd probably want a some_container<std::vector<FeatureItem>> where the inner std::vector<> is limited to batches of let's say 1 million features to avoid too big memory reallocations.

`std::deque` might be an alternative to `std::vector`.

Reply:
> make_shared would improve allocations of std::shared.

This is what is used in master currently

> std::deque might be an alternative to std::vector.

I've just tested it https://github.com/OSGeo/gdal/commit/83ec3c62f125034c146c7bd57e68b312bee66a79 . 
RAM usage decrease a bit compared to the current state of this PR. Execution time slightly increases however, but remains faster than master

```
$ /usr/bin/time --verbose ogr2ogr out.fgb nz-building-outlines.fgb
	Command being timed: "ogr2ogr out.fgb nz-building-outlines.fgb"
	User time (seconds): 20.37
	System time (seconds): 3.13
	Percent of CPU this job got: 99%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:23.52
	Average shared text size (kbytes): 0
	Average unshared data size (kbytes): 0
	Average stack size (kbytes): 0
	Average total size (kbytes): 0
	Maximum resident set size (kbytes): 365436
	Average resident set size (kbytes): 0
	Major (requiring I/O) page faults: 0
	Minor (reclaiming a frame) page faults: 112427
	Voluntary context switches: 233
	Involuntary context switches: 565
	Swaps: 0
	File system inputs: 0
	File system outputs: 6857352
	Socket messages sent: 0
	Socket messages received: 0
	Signals delivered: 0
	Page size (bytes): 4096
	Exit status: 0
```

```
VmPeak:	  472520 kB
VmSize:	  444724 kB
VmLck:	       0 kB
VmPin:	       0 kB
VmHWM:	  365616 kB
VmRSS:	  336872 kB
RssAnon:	  304476 kB
RssFile:	   32396 kB
RssShmem:	       0 kB
VmData:	  306300 kB
VmStk:	     136 kB
VmExe:	      12 kB
VmLib:	   65524 kB
VmPTE:	     824 kB
VmSwap:	       0 kB
```

This might be a good compromise for all use cases as it doesn't involve the occasional big reallocation pattern of std::vector when its capacity is exhausted.
@bjornharrtell what do you think ? Should we go with https://github.com/OSGeo/gdal/commit/83ec3c62f125034c146c7bd57e68b312bee66a79  on top of this PR ?

Reply:
Sure, RAM is precious. 🙂

Reply:
https://github.com/OSGeo/gdal/commit/83ec3c62f125034c146c7bd57e68b312bee66a79 merged into this PR